Mathlib/Algebra/Ring/Associator.lean

```diff
@@ -11,14 +11,61 @@ import Mathlib.Tactic.Abel
 /-!
 ## Associator in a ring
 If `R` is a non-associative ring, then  `(x * y) * z - x * (y * z)` is called the `associator` of
-ring elements `x y z : R`. We realize the associator as a map that is an `AddMonoidHom` in each
-argument. The associator vanishes exactly when `R` is associative. We prove a similar statement for
-semirings by directly comparing the maps `mulLeft₃` and `mulRight₃`, the `AddMonoidHom` versions of
-the multiplications `(x * y) * z` and `x * (y * z)`.
+ring elements `x y z : R`.
+
+The associator vanishes exactly when `R` is associative. We prove a similar statement for the
+`AddMonoidHom` bundled version of the associator, as well as for semirings by directly comparing the
+maps `mulLeft₃` and `mulRight₃`, the multiplications `(x * y) * z` and `x * (y * z)`, and for their
+`AddMonoidHom` versions.
 -/
 
 universe u
 
+namespace Mul
+variable {R : Type u} [Mul R]
+
+/-- The multiplication `(x * y) * z` -/
+def mulLeft₃ (x y z : R) := (x * y) * z
+
+/-- The multiplication `x * (y * z)` -/
+def mulRight₃ (x y z : R) := x * (y * z)
+
+@[simp]
+theorem mulLeft₃_apply (x y z : R) : mulLeft₃ x y z = (x * y) * z := rfl
+
+@[simp]
+theorem mulRight₃_apply (x y z : R) : mulRight₃ x y z = x * (y * z) := rfl
+
+/-- The multiplications `(x * y) * z` and `x * (y * z)` agree iff multiplication is associative. -/
+theorem mulLeft₃_eq_mulRight₃_iff_associative :
+    mulLeft₃ (R := R) = mulRight₃ ↔ Std.Associative (fun (x y : R) ↦ x * y) where
+  mp h := ⟨fun x y z ↦ by rw [← mulLeft₃_apply, ← mulRight₃_apply, h]⟩
+  mpr h := by ext x y z; simp [Std.Associative.assoc]
+end Mul
+
+namespace NonUnitalNonAssocRing
+variable {R : Type u} [NonUnitalNonAssocRing R]
+/-- The associator `(x * y) * z - x * (y * z)` -/
+def associator (x y z : R) := (x * y) * z - x * (y * z)
+
+theorem associator_eq_mulLeft₃_sub_mulRight₃ :
+    associator = Mul.mulLeft₃ (R := R) - Mul.mulRight₃ := by
+  funext x y z
+  simp [associator]
+
+theorem associative_iff_associator_eq_zero : Std.Associative (fun (x y : R) ↦ x * y) ↔
+    associator (R := R) = 0 := by
```

Comment:
Can you please start multi-line return types on a new line? Currently, it looks like you are proving `associator (R := R) = 0`!
```suggestion
theorem associative_iff_associator_eq_zero :
    Std.Associative (fun (x y : R) ↦ x * y) ↔ associator (R := R) = 0 := by
```